cli/src/semgrep/project.py

```diff
@@ -24,3 +38,71 @@ def get_project_url() -> Optional[str]:
         except Exception as e:
             logger.debug(f"Failed to get project url from .git/config: {e}")
             return None
+
+
+@define
+class ProjectConfig:
+    metadata: Dict[str, Any] = Factory(dict)
+
+    @staticmethod
+    def is_project_config_file(file_path: Path) -> bool:
+        return CONFIG_FILE_PATTERN.search(file_path.name) is not None
+
+    @classmethod
+    def _find_all_config_files(cls, src_directory: Path, cwd_path: Path) -> List[Path]:
+        conf_files = []
+
+        # Populate stack of directories to traverse
+        stack = {cwd_path}
+        temp_path = src_directory
+        dir_route = cwd_path.relative_to(src_directory)
+        for parent in dir_route.parents:
+            temp_path = temp_path / parent
+            stack.add(temp_path)
+
+        # Traverse stack looking for config files
+        while stack:
+            cur_path = stack.pop()
+            conf_files += [
+                f for f in cur_path.iterdir() if cls.is_project_config_file(f)
+            ]
+        return conf_files
+
+    @classmethod
+    def load_from_file(cls, file_path: Path) -> Dict[str, Any]:
+        yaml = ruamel.yaml.YAML(typ="safe")
+        logger.debug(f"Loading semgrepconfig file: {file_path}")
+        with file_path.open("r") as fp:
+            config: Dict[str, Any] = yaml.load(fp)
+            return config
+
+    @classmethod
+    def load_all(cls) -> "ProjectConfig":
+        src_directory = get_git_root_path()
+        cwd_path = Path.cwd()
+        conf_files = cls._find_all_config_files(src_directory, cwd_path)
+
+        # Sort by depth asc
+        conf_files.sort(key=lambda x: len(x.parts))
+
+        # Merge metadata from all config files
+        all_metadata: Dict[Any, Any] = {}
+        for conf_file in conf_files:
+            local_metadata = cls.load_from_file(conf_file)
+            all_metadata = {**all_metadata, **local_metadata}
+        return cls(all_metadata)
+
+    def to_dict(self) -> Dict[str, Any]:
+        # Convert all tag values to strings
+        data = deepcopy(self.metadata)
+        tags = data.pop("tags", None)
+        if tags:
+            data["tags"] = []
+            for tag in tags:
+                if isinstance(tag, dict):
+                    data["tags"].append(json.dumps(tag))
```

Comment:
Might be nice to call out explicitly (with examples?) the types of "tags" data you are supporting here and why you are flattening them all to be of type string. It strikes me as a little odd that we support standalone-string tags and also dictionary-type tags. Can we simplify and just support one or has that ship sailed?

Reply:
From a user perspective, when editing config files, it would be clearer how data will be used/what is expected of me with a dict. For example, if I received a config template with something like this, I could run semgrep without ever adding the correct config data:

```
{
    'name=',
    'animal=',
    'age='
}
```

Just by nature of the parser - I wouldn't be able to do the same with a config template like this (which is a good thing!):

```
{
    'name':,
    'animal':,
    'age':
}
```

Reply:
The app only supports string types today. The ProjectConfig yaml intentionally supports both string and dict types for the tags for forward-compatiblity. But for the sake of simplicity, we could limit the ProjectConfig tags format to an array of strings.